middle_end/flambda/invariant_params.ml

```diff
@@ -65,47 +65,34 @@ let implies relation from to_ =
       relation
 
 let transitive_closure state =
-  let union s1 s2 =
-    match s1, s2 with
-    | Top, _ | _, Top -> Top
-    | Implication s1, Implication s2 ->
-      Implication (Variable.Pair.Set.union s1 s2)
+  (* Depth-first search for all implications for one argument.
+     Arguments are moved from candidate to frontier, assuming
+     they are newly added to the result. *)
+  let rec loop candidate frontier result =
+    match (candidate, frontier) with
+    | ([], []) -> Implication result
+    | ([], frontier::fs) ->
+      (* Obtain fresh candidate for the frontier argument. *)
+      (match (try Variable.Pair.Map.find frontier state with
+              | Not_found -> Implication Variable.Pair.Set.empty) with
+       | Top -> Top
+       | Implication candidate ->
+         loop (Variable.Pair.Set.elements candidate) fs result)
+    | (candidate::cs, frontier) ->
+      let result' = Variable.Pair.Set.add candidate result in
+      if result' != result then
+        (* Result change means candidate becomes part of frontier. *)
+        loop cs (candidate :: frontier) result'
+      else
+        loop cs frontier result
   in
-  let equal s1 s2 =
-    match s1, s2 with
-    | Top, Implication _ | Implication _, Top -> false
-    | Top, Top -> true
-    | Implication s1, Implication s2 -> Variable.Pair.Set.equal s1 s2
-  in
-  let update arg state =
-    let original_set =
-      try Variable.Pair.Map.find arg state with
-      | Not_found -> Implication Variable.Pair.Set.empty
-    in
-    match original_set with
-    | Top -> state
-    | Implication arguments ->
-        let set =
-          Variable.Pair.Set.fold
-            (fun orig acc->
-               let set =
-                 try Variable.Pair.Map.find orig state with
-                 | Not_found -> Implication Variable.Pair.Set.empty in
-               union set acc)
-            arguments original_set
-        in
-        Variable.Pair.Map.add arg set state
-  in
-  let once state =
-    Variable.Pair.Map.fold (fun arg _ state -> update arg state) state state
-  in
-  let rec fp state =
-    let state' = once state in
-    if Variable.Pair.Map.equal equal state state'
-    then state
-    else fp state'
-  in
-  fp state
+    Variable.Pair.Map.filter_map
```

Comment:
You use `filter_map` but I don't see any actual filtering being done below, only a `Some` case. (This may be a leftover from a version that tried to trim Top elements?)